File: src/main/GO/leetcode/lists/hot100/LeetCode114.go
package hot100

func flatten(root *TreeNode) {

	if root == nil {
		return
	}

	stack := make([]*TreeNode, 0)
	stack = append(stack, root)
	for len(stack) > 0 {
		cur := stack[len(stack)-1]
		stack = stack[0 : len(stack)-1]
		if cur.Right != nil {
			stack = append(stack, cur.Right)
		}
		if cur.Left != nil {
			stack = append(stack, cur.Left)
		}
		if len(stack) > 0 {
			cur.Right = stack[len(stack)-1]
			stack = stack[0 : len(stack)-1]
		}
		cur.Left = nil
	}
}

File: src/main/java/leetcode/problems/LeetCode114.java
package leetcode.problems;

/**
 * @author wujingxinit@outlook.com
 * @date 2023/7/12 00:10
 */
public class LeetCode114 {

    public void flatten(TreeNode root) {
        // base case
        if (root == null) {
            return;
        }

        // 利用定义，把左右子树拉平
        flatten(root.left);
        flatten(root.right);

        /* 后序遍历位置 */
        // 1、左右子树已经被拉平成一条链表
        TreeNode left = root.left;
        TreeNode right = root.right;

        // 2、将左子树作为右子树
        root.left = null;
        root.right = left;

        // 3、将原先的右子树接到当前右子树的末端
        TreeNode p = root;
        while (p.right != null) {
            p = p.right;
        }
        p.right = right;

    }
}
